File: MatlabOctaveCodeSnippets/snip_systems_recorded_noise.m
[x,Fs,b]=readwav('filteredNoise.wav'); %recorded output
N=1024; %number of FFT points
M=floor(length(x)/N); %number of segments of N samples each
x=x(1:N*M); x=x(:); %make sure x is a column vector with N*M samples
xsegments=reshape(x,N,M); %segment x into M blocks
X=abs(fft(xsegments)); %obtain the magnitude for each segment
X=mean(transpose(X)); %transpose: the mean has to be over the FFTs
f=Fs/N*(0:N/2)/1000; %create abscissa in kHz. Fs/N is the bin width
plot(f,20*log10(X(1:N/2+1))),xlabel('f (kHz)'),ylabel('|H(f)| (dBW)')

File: PythonCodeSnippets/snip_systems_recorded_noise.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile

def readwav(filename):
    """Função de leitura de arquivo WAV simplificada."""
    return wavfile.read(filename)

x, Fs = readwav('filteredNoise.wav')

N = 1024

M = len(x) // N

x = x[:N*M].reshape(-1, 1)

xsegments = x.reshape(N, M)

X = np.abs(np.fft.fft(xsegments, axis=0))

X = np.mean(X.T, axis=0)

f = Fs / N * np.arange(N/2 + 1) / 1000

plt.plot(f, 20 * np.log10(X[:N//2 + 1]))
plt.xlabel('f (kHz)')
plt.ylabel('|H(f)| (dBW)')
plt.show()
